Public/Src/Engine/Dll/EngineSchedule.cs

```diff
@@ -771,21 +771,46 @@ private bool ShouldSerializeOptimizationDataStructurePostExecution()
                 outputDirectories = scheduler.PipGraph.AllDirectoriesContainingOutputs().Select(d => d.ToString(scheduler.Context.PathTable)).ToList();
             }
 
+            var scrubber = new DirectoryScrubber(
+                cancellationToken: scheduler.Context.CancellationToken,
+                loggingContext: loggingContext,
+                loggingConfiguration: configuration.Logging,
+                maxDegreeParallelism: Environment.ProcessorCount,
+                tempDirectoryCleaner: tempCleaner);
+
+            var journalFiles = SharedOpaqueJournal
+                .FindAllJournalFiles(configuration.Layout.SharedOpaqueJournalDirectory.ToString(scheduler.Context.PathTable))
+                .ToArray();
+            var distinctRecordedWrites = journalFiles
+                .AsParallel()
+                .WithDegreeOfParallelism(Environment.ProcessorCount)
+                .WithCancellation(scheduler.Context.CancellationToken)
+                .SelectMany(SharedOpaqueJournal.ReadRecordedWritesFromJournalWrapExceptions)
+                .Distinct(StringComparer.OrdinalIgnoreCase)

```

Comment:
>Distinct(StringComparer.OrdinalIgnoreCase) [](start = 17, length = 42)

Do we actually need Distinct() here?
My understanding is that duplicates can only occur when a double write happened (and the corresponding build failed). So it doesn't look like the most common case. An actual race trying to delete the same file is not a problem AFAIK. On the other hand, a Distinct() among lots of files could be time consuming

Reply:
A path is written to a journal as soon as a Write operation is reported by the sandbox.  The same process can write multiple times to the same file, so yes, we end up having a ton of duplicates here.

Reply:
I'm not sure about the perf implications and the behavior under race conditions.  Judging by a few examples I've seen, the number of entries in a journal can easily be 5x the number of distinct paths; calling File.Delete for each entry doesn't sound beneficial to me.  Also conceptually, if we don't deduplicate here, we cannot report how many files we actually wanted to delete and how many we actually deleted.

Reply:
That 5X surprises me... shared opaque outputs are enforced to be uniquely produced. So a duplicate file means a build break.

---
In reply to: [334131960](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334131960) [](ancestors = 334131960)

Reply:
And following the assumption that duplicate files are not common, I don't think reporting more files in that case is a big problem

---
In reply to: [334138336](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334138336) [](ancestors = 334138336,334131960)

Reply:
Each shared opaque output is indeed uniquely produced by a single pip; but that one pip can write to it as many times as it wants, and every time it writes to it we get a report about it from the sandbox.  The journal records reports from the sandbox as soon as they are received, hence duplication.

Reply:
Ah, you are correct, yes. You could consider keeping a set of already written paths and check before writing again. We are already keeping similar sets/maps as such, from what I could see.

---
In reply to: [334130813](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334130813) [](ancestors = 334130813)

Reply:
That would also reduce the number of writes/size of those files (which from what you are seeing seems to be 5X what they could be)

---
In reply to: [334141163](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334141163) [](ancestors = 334141163,334130813)

Reply:
made SharedOpaqueJournal dedupe paths on the fly, before writing them to disk